File: src/eQuantic.UI.Runtime/src/shared/components/CodeDiffLayout.ts
import { $eq, CodeCollapse, CodeDiffFold, CodeFiller, CodeLineChange, CodeRows } from "../runtime-exports";

export class CodeDiffLayout { declare original: CodeRows; declare modified: CodeRows; declare folds: CodeDiffFold[]; constructor(original: any = null, modified: any = null, folds: any = null) { this.original = original; this.modified = modified; this.folds = folds; } equals(o: unknown) { return o instanceof CodeDiffLayout && $eq.equals(this.original, o.original) && $eq.equals(this.modified, o.modified) && $eq.equals(this.folds, o.folds); } with(patch: any) { return new CodeDiffLayout(('original' in patch ? patch.original : this.original), ('modified' in patch ? patch.modified : this.modified), ('folds' in patch ? patch.folds : this.folds)); } static sideBySide(changes: any, originalLines: number, modifiedLines: number, context: number, expanded: any, gaps: any, foldLabel: ((value: number) => string) | null) { let originalFillers: CodeFiller[] = [];
    let modifiedFillers: CodeFiller[] = [];
    let gap = 0;
    for (const change of changes) {
        gap = CodeDiffLayout.addGapsBefore(change, gaps, gap, originalFillers, modifiedFillers);
        let difference = change.modifiedCount - change.originalCount;
        if (difference > 0) originalFillers.push(new CodeFiller(change.originalStart + change.originalCount, difference)); else if (difference < 0) modifiedFillers.push(new CodeFiller(change.modifiedStart + change.modifiedCount, -difference));
    }
    CodeDiffLayout.addGapsBefore(null, gaps, gap, originalFillers, modifiedFillers);
    let [originalRuns, modifiedRuns, folds] = CodeDiffLayout.unchangedRuns(changes, originalLines, modifiedLines, context, expanded, gaps, foldLabel);
    return new CodeDiffLayout(new CodeRows(originalLines, originalFillers, originalRuns), new CodeRows(modifiedLines, modifiedFillers, modifiedRuns), folds); } static inline(changes: any, originalLines: number, modifiedLines: number, context: number, expanded: any, gaps: any, foldLabel: ((value: number) => string) | null) { let originalGaps: CodeFiller[] = [];
    let removed: CodeFiller[] = [];
    let gap = 0;
    for (const change of changes) {
        gap = CodeDiffLayout.addGapsBefore(change, gaps, gap, originalGaps, removed);
        if (change.originalCount > 0) removed.push(new CodeFiller(change.modifiedStart, change.originalCount, change.originalStart));
    }
    CodeDiffLayout.addGapsBefore(null, gaps, gap, originalGaps, removed);
    let [originalRuns, modifiedRuns, folds] = CodeDiffLayout.unchangedRuns(changes, originalLines, modifiedLines, context, expanded, gaps, foldLabel);
    return new CodeDiffLayout(new CodeRows(originalLines, originalGaps, originalRuns), new CodeRows(modifiedLines, removed, modifiedRuns), folds); } static addGapsBefore(change: CodeLineChange | null, gaps: any, next: number, original: CodeFiller[], modified: CodeFiller[]) { if (gaps == null) return next;
    while (next < gaps.length && (change == null || gaps[next].originalLine <= change.originalStart && gaps[next].modifiedLine <= change.modifiedStart)) {
        original.push(new CodeFiller(gaps[next].originalLine, 1, undefined, gaps[next].header));
        modified.push(new CodeFiller(gaps[next].modifiedLine, 1, undefined, gaps[next].header));
        next++;
    }
    return next; } static unchangedRuns(changes: any, originalLines: number, modifiedLines: number, context: number, expanded: any, gaps: any, foldLabel: ((value: number) => string) | null): [CodeCollapse[], CodeCollapse[], CodeDiffFold[]] { let original: CodeCollapse[] = [];
    let modified: CodeCollapse[] = [];
    let folds: CodeDiffFold[] = [];
    let originalAt = 0;
    let modifiedAt = 0;
    for (let i = 0; i <= changes.length; i++) {
        let originalEnd = i < changes.length ? changes[i].originalStart : originalLines;
        let modifiedEnd = i < changes.length ? changes[i].modifiedStart : modifiedLines;
        let length = Math.min(originalEnd - originalAt, modifiedEnd - modifiedAt);
        let before = i > 0 ? context : 0;
        let after = i < changes.length ? context : 0;
        let hidden = length - before - after;
        if (hidden >= CodeDiffLayout.fewestFolded && !((($r) => $r == null ? null : $eq.collections.contains($r, originalAt + before))(expanded) ?? false) && !CodeDiffLayout.crossesAGap(gaps, originalAt + before, originalAt + before + hidden)) {
            let label = foldLabel?.(hidden);
            original.push(new CodeCollapse(originalAt + before, originalAt + before + hidden - 1, undefined, label));
            modified.push(new CodeCollapse(modifiedAt + before, modifiedAt + before + hidden - 1, undefined, label));
            folds.push(new CodeDiffFold(originalAt + before, modifiedAt + before, hidden));
        }
        if (i < changes.length) {
            originalAt = changes[i].originalStart + changes[i].originalCount;
            modifiedAt = changes[i].modifiedStart + changes[i].modifiedCount;
        }
    }
    return [original, modified, folds]; } foldOfModified(line: number) { for (const fold of this.folds) if (line >= fold.modifiedLine && line < fold.modifiedLine + fold.count) return fold;
    return null; } foldOfOriginal(line: number) { for (const fold of this.folds) if (line >= fold.originalLine && line < fold.originalLine + fold.count) return fold;
    return null; } static crossesAGap(gaps: any, from: number, to: number) { if (gaps == null) return false;
    for (const gap of gaps) if (gap.originalLine > from && gap.originalLine <= to) return true;
    return false; } static defaultContext = 3; static fewestFolded = 3; toString() { return `CodeDiffLayout { Original = ${this.original}, Modified = ${this.modified}, Folds = ${this.folds} }`; } }

File: src/eQuantic.UI.Code/Diff/CodeDiffLayout.cs
namespace eQuantic.UI.Code;

/// <summary>
/// The rows a diff is drawn on (docs/CODE-EDITOR-PLAN.md, the shape, §9): for the two sides of a
/// side-by-side view, <see cref="Original"/> and <see cref="Modified"/>, which take the same number
/// of rows and keep every change level; for an inline view, <see cref="Modified"/> alone, with the
/// lines each change removed drawn before the lines that replaced them.
/// <para>
/// Runs of unchanged lines longer than the context either side of a change can use are folded into
/// one row, on both sides at once, unless the view has opened them: both factories take the runs
/// opened, named by the first line each hides on the original side, which a change elsewhere does
/// not move.
/// </para>
/// </summary>
public sealed record CodeDiffLayout(CodeRows Original, CodeRows Modified)
{
    /// <summary>How many unchanged lines stay in view either side of a change.</summary>
    public const int DefaultContext = 3;

    /// <summary>A run is folded only when it hides at least this many lines: folding two lines
    /// into one row saves a row and costs a press.</summary>
    private const int FewestFolded = 3;

    /// <summary>
    /// The rows of both sides of a side-by-side view: at each change the side with fewer lines is
    /// padded to the other's, AFTER its own lines, so the two stay level; and every long unchanged
    /// run is folded on both.
    /// </summary>
    public static CodeDiffLayout SideBySide(IReadOnlyList<CodeLineChange> changes, int originalLines,
        int modifiedLines, int context = DefaultContext, IReadOnlyCollection<int>? expanded = null)
    {
        var originalFillers = new List<CodeFiller>();
        var modifiedFillers = new List<CodeFiller>();
        foreach (var change in changes)
        {
            var difference = change.ModifiedCount - change.OriginalCount;
            if (difference > 0)
                originalFillers.Add(new CodeFiller(change.OriginalStart + change.OriginalCount, difference));
            else if (difference < 0)
                modifiedFillers.Add(new CodeFiller(change.ModifiedStart + change.ModifiedCount, -difference));
        }
        var (originalRuns, modifiedRuns) = UnchangedRuns(changes, originalLines, modifiedLines, context, expanded);
        return new CodeDiffLayout(
            new CodeRows(originalLines, originalFillers, originalRuns),
            new CodeRows(modifiedLines, modifiedFillers, modifiedRuns));
    }

    /// <summary>
    /// The rows of an inline view, on the modified text: before the lines a change added, the lines
    /// it removed, drawn from the original as fillers; and every long unchanged run folded.
    /// <see cref="Original"/> is the original's own rows, with the same runs folded, for a view that
    /// shows the original side too.
    /// </summary>
    public static CodeDiffLayout Inline(IReadOnlyList<CodeLineChange> changes, int originalLines,
        int modifiedLines, int context = DefaultContext, IReadOnlyCollection<int>? expanded = null)
    {
        var removed = new List<CodeFiller>();
        foreach (var change in changes)
        {
            if (change.OriginalCount > 0)
                removed.Add(new CodeFiller(change.ModifiedStart, change.OriginalCount, change.OriginalStart));
        }
        var (originalRuns, modifiedRuns) = UnchangedRuns(changes, originalLines, modifiedLines, context, expanded);
        return new CodeDiffLayout(
            new CodeRows(originalLines, [], originalRuns),
            new CodeRows(modifiedLines, removed, modifiedRuns));
    }

    /// <summary>
    /// The unchanged runs to fold on each side: between two changes (or a change and an end of the
    /// file), everything but <paramref name="context"/> lines next to each change, when that leaves at
    /// least <see cref="FewestFolded"/> lines to hide and the view has not opened the run.
    /// </summary>
    private static (List<CodeCollapse> Original, List<CodeCollapse> Modified) UnchangedRuns(
        IReadOnlyList<CodeLineChange> changes, int originalLines, int modifiedLines, int context,
        IReadOnlyCollection<int>? expanded)
    {
        var original = new List<CodeCollapse>();
        var modified = new List<CodeCollapse>();
        var originalAt = 0;
        var modifiedAt = 0;
        for (var i = 0; i <= changes.Count; i++)
        {
            // The unchanged run from where the previous change ended to where this one starts (or
            // to the end of the file after the last change). The two sides share its length.
            var originalEnd = i < changes.Count ? changes[i].OriginalStart : originalLines;
            var modifiedEnd = i < changes.Count ? changes[i].ModifiedStart : modifiedLines;
            var length = Math.Min(originalEnd - originalAt, modifiedEnd - modifiedAt);
            // The file's own edges need no context: nothing is there to read it against.
            var before = i > 0 ? context : 0;
            var after = i < changes.Count ? context : 0;
            var hidden = length - before - after;
            if (hidden >= FewestFolded && !(expanded?.Contains(originalAt + before) ?? false))
            {
                original.Add(new CodeCollapse(originalAt + before, originalAt + before + hidden - 1));
                modified.Add(new CodeCollapse(modifiedAt + before, modifiedAt + before + hidden - 1));
            }
            if (i < changes.Count)
            {
                originalAt = changes[i].OriginalStart + changes[i].OriginalCount;
                modifiedAt = changes[i].ModifiedStart + changes[i].ModifiedCount;
            }
        }
        return (original, modified);
    }
}
